File: bindings/python/demo/07_dmd.py
# Copyright (c) 2026 Edison Manrique Chocce
# Todos los derechos reservados.
# Licencia Dual: AGPL-3.0 | Comercial según LICENSE.
"""
demo/07_dmd.py — Dynamic Mode Decomposition (DMD)
Extrae frecuencias dominantes de sistemas espacio-temporales.
"""
import sys
if sys.stdout.encoding != "utf-8":
    sys.stdout.reconfigure(encoding="utf-8")

import math
sys.path.insert(0, "..")
from ermc import DMD

print("=" * 60)
print("  ERMC | Demo 07: DMD - Modos Dinamicos Dominantes")
print("=" * 60)

# --- Experimento 1: Señal mono-frecuencia conocida ---
print("\n[1] Senal mono-frecuencia (omega = pi rad/s):")
dt = 0.05
n = 60
omega_true = math.pi
snapshots_1 = [
    [math.sin(omega_true * k * dt) for k in range(n)],
    [math.cos(omega_true * k * dt) for k in range(n)],
]
freq_1 = DMD.dominant_frequency(snapshots_1, dt)
error_1 = abs(freq_1 - omega_true)
print(f"  omega teorico: {omega_true:.6f} rad/s")
print(f"  omega DMD:     {freq_1:.6f} rad/s")
print(f"  Error:         {error_1:.2e}")
print(f"  {'[PASS]' if error_1 < 0.05 else '[FAIL]'}")

# --- Experimento 2: Batimiento de dos frecuencias ---
print("\n[2] Batimiento: suma de dos frecuencias (omega1=2, omega2=7):")
omega1, omega2 = 2.0, 7.0
dt2 = 0.02
n2 = 80
snapshots_2 = [
    [math.sin(omega1 * k * dt2) + 0.5 * math.sin(omega2 * k * dt2) for k in range(n2)],
    [math.cos(omega1 * k * dt2) + 0.5 * math.cos(omega2 * k * dt2) for k in range(n2)],
]
freq_2 = DMD.dominant_frequency(snapshots_2, dt2)
print(f"  Frecuencia dominante extraida: {freq_2:.4f} rad/s")
print(f"  (La componente mas energetica es omega1 = {omega1} rad/s)")

# --- Experimento 3: Sistema fisico — oscilador de Van der Pol aproximado ---
print("\n[3] Datos de tipo Van der Pol (oscilador de ciclo limite):")
# Simulacion simplificada de trayectoria de VdP
eps = 0.1
x, y = 0.1, 0.0
dt3 = 0.02
n3 = 100
trj_x, trj_y = [], []
for _ in range(n3):
    dx = y
    dy = eps * (1 - x * x) * y - x
    x += dx * dt3
    y += dy * dt3
    trj_x.append(x)
    trj_y.append(y)

freq_vdp = DMD.dominant_frequency([trj_x, trj_y], dt3)
print(f"  Frecuencia de ciclo limite detectada: {freq_vdp:.4f} rad/s")
print(f"  (Esperado aprox: 1.0 rad/s para eps pequeño)")

print("\n[OK] Demo 07 completado.\n")

File: bindings/bun/demo/07_dmd.ts
/**
 * Copyright (c) 2026 Edison Manrique Chocce
 * Licencia Dual: AGPL-3.0 | Comercial según LICENSE.
 *
 * demo/07_dmd.ts — Dynamic Mode Decomposition
 */
import { DMD } from "../src/index";

console.log("=".repeat(60));
console.log("  ERMC | Demo 07: DMD - Modos Dinámicos Dominantes");
console.log("=".repeat(60));

// [1] Mono-frecuencia conocida
const omegaTrue = Math.PI;
const dt1 = 0.05, n1 = 60;
const snaps1 = [
  Array.from({ length: n1 }, (_, k) => Math.sin(omegaTrue * k * dt1)),
  Array.from({ length: n1 }, (_, k) => Math.cos(omegaTrue * k * dt1)),
];
const freq1 = DMD.dominantFrequency(snaps1, dt1);
const err1 = Math.abs(freq1 - omegaTrue);
console.log("\n[1] Señal mono-frecuencia (omega = pi rad/s):");
console.log(`  omega teórico: ${omegaTrue.toFixed(6)} rad/s`);
console.log(`  omega DMD:     ${freq1.toFixed(6)} rad/s`);
console.log(`  Error:         ${err1.toExponential(2)}`);
console.log(`  ${err1 < 0.05 ? "[PASS]" : "[FAIL]"}`);

// [2] Batimiento de dos frecuencias
const omega1 = 2.0, omega2 = 7.0;
const dt2 = 0.02, n2 = 80;
const snaps2 = [
  Array.from({ length: n2 }, (_, k) => Math.sin(omega1 * k * dt2) + 0.5 * Math.sin(omega2 * k * dt2)),
  Array.from({ length: n2 }, (_, k) => Math.cos(omega1 * k * dt2) + 0.5 * Math.cos(omega2 * k * dt2)),
];
const freq2 = DMD.dominantFrequency(snaps2, dt2);
console.log("\n[2] Batimiento (omega1=2, omega2=7):");
console.log(`  Frecuencia dominante extraída: ${freq2.toFixed(4)} rad/s`);
console.log(`  (Componente más energética = omega1 = ${omega1} rad/s)`);

// [3] Van der Pol
let x = 0.1, y = 0.0;
const dt3 = 0.02, n3 = 100, eps = 0.1;
const trjX: number[] = [], trjY: number[] = [];
for (let _ = 0; _ < n3; _++) {
  const dx = y;
  const dy = eps * (1 - x * x) * y - x;
  x += dx * dt3;
  y += dy * dt3;
  trjX.push(x);
  trjY.push(y);
}
const freqVdp = DMD.dominantFrequency([trjX, trjY], dt3);
console.log("\n[3] Oscilador de Van der Pol (ciclo límite):");
console.log(`  Frecuencia detectada: ${freqVdp.toFixed(4)} rad/s`);
console.log(`  (Esperado aprox: 1.0 rad/s para eps pequeño)`);

console.log("\n[OK] Demo 07 completado.\n");
